### PR TITLE
Fix bootloader advanced use

### PR DESCRIPTION
**Pull request by otavio**
We found some corner cases which are raised when external layers do make some fancy use of the bootloader settings. This is sometimes required and this is very hard to debug. We ended working on the required fixes which covered the known cases.

### CODE REVIEW

**Comment by otavio** [comment]
@thochstein @zandrey @angolini could you take a look at this?

**Comment by angolini** [comment]
is this related with #53 somehow?

**Review by angolini — APPROVED** [review]

**Comment by angolini** [comment]
Great PR, @otavio !
I'm not sure if I can test it today, so please don't wait for me. But I if I face something I let you know

**Comment by otavio** [comment]
@angolini #53 seems to be u-boot-imx related. Not sure this is the same issue.

File: conf/machine/include/imx-base.inc [hunk 1]
```diff
@@ -63,6 +63,10 @@ UBOOT_BINARY ?= "u-boot.${UBOOT_SUFFIX}"
 IMX_DEFAULT_BOOTLOADER ??= "u-boot-fslc"
 UBOOT_SUFFIX ?= "${UBOOT_SUFFIX_pn-${IMX_DEFAULT_BOOTLOADER}}"
 
+# We need to export the original variable to allow it to be used when generating
+# wic based images.
+SPL_BINARY ?= "${SPL_BINARY_pn-${IMX_DEFAULT_BOOTLOADER}}"
+
```
Comment (thochstein) [inline conf/machine/include/imx-base.inc:69]: Was there a build error? Can you mention in the commit message how the problem manifested itself?

File: recipes-bsp/u-boot/libubootenv_%.bbappend [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+# This is not intended to be permanent but we need to get the integration
+# working and there is no good solution for now so we are adding this in a
+# non-intrusive way and using the `IMX_DEFAULT_BOOTLOADER` as a guard to do any
+# code execution.
+
+def fixup_uboot_config_dependency(d):
+    ubootmachine = d.getVar("UBOOT_MACHINE")
+    ubootconfig = (d.getVar('UBOOT_CONFIG') or "").split()
+    imx_default_bootloader = d.get('IMX_DEFAULT_BOOTLOADER')
+
```
Comment (thochstein) [inline recipes-bsp/u-boot/libubootenv_%.bbappend:12]: Can we not do similar as for `SPL_BINARY`?

```
UBOOT_MACHINE ?= "${UBOOT_MACHINE_pn-${IMX_DEFAULT_BOOTLOADER}}"
```

File: conf/machine/include/imx-base.inc (hunk 1 above)
Comment (otavio) [reply conf/machine/include/imx-base.inc:69]: We couldn't reference it on the wks file. It took me some time to find the root cause as the error wan't very useful.

File: recipes-bsp/u-boot/libubootenv_%.bbappend (hunk 2 above)
Comment (otavio) [reply recipes-bsp/u-boot/libubootenv_%.bbappend:12]: No; the `uboot-config` fails if both are set ("" is set).

**Review by thochstein — APPROVED** [review]
Looks good to me.

**Comment by thochstein** [comment]
This breaks `u-boot` for `fsl-xwayland/imx6sllevk`.

```
| install: cannot stat '/opt/work/upstream/fsl-xwayland/tmp/work/imx6sllevk-fsl-linux-gnueabi/u-boot-fslc/v2020.10+gitAUTOINC+80c23498b4-r0/build/mx6sllevk_config/u-boot-imx': No such file or directory
```
 
It appears that `SPL_BINARY` evaluates to the package name. I had similar issue #704 in layer `meta-imx` for `UBOOT_SUFFIX` and `UBOOT_MAKE_TARGET`, yet those variables don't show a problem here.

```
nxa21648@lsv07069:/opt/work/upstream/fsl-xwayland$ MACHINE=imx6sllevk bitbake u-boot -e | grep ^SPL_BINARY
SPL_BINARY="\${SPL_BINARY_pn-u-boot-imx}"
SPL_BINARYNAME="\${SPL_BINARY_pn-u-boot-imx}"
```

**Pull request merged**